src/Fusonic/Linq/Iterator/OrderIterator.php

```diff
@@ -71,43 +69,50 @@ public function orderItems()
             return;
         }
 
-        $firstOrderKey = $orderKeyFunc($itemIterator->current());
+        $this->orderedIterator = $this->iterator;
 
-        $sortType = Helper\LinqHelper::LINQ_ORDER_TYPE_NUMERIC;
+        // Ugly hack for PHP 5.3 as the calling context is not handled correctly in anonymous functions
+        $self = $this;
+        $orderFuncs = $this->orderFuncs;
 
-        if ($firstOrderKey instanceof \DateTime) {
-            $sortType = Helper\LinqHelper::LINQ_ORDER_TYPE_DATETIME;
-        } elseif (!is_numeric($firstOrderKey)) {
-            $sortType = Helper\LinqHelper::LINQ_ORDER_TYPE_ALPHANUMERIC;
-        }
+        $this->orderedIterator->uasort(function($a, $b) use ($self, $orderFuncs) {
+            $result = 0;
+            foreach ($orderFuncs as $orderFunc) {
+                $func = $orderFunc['func'];
 
-        $keyMap = array();
-        $valueMap = array();
+                if ($orderFunc['direction'] === Helper\LinqHelper::LINQ_ORDER_ASC) {
+                    $result = $self->compare($func($a), $func($b));
+                } else {
+                    $result = $self->compare($func($b), $func($a));
+                }
 
-        foreach ($itemIterator as $value) {
-            $orderKey = $orderKeyFunc != null ? $orderKeyFunc($value) : $value;
-            if ($sortType == Helper\LinqHelper::LINQ_ORDER_TYPE_DATETIME) {
-                $orderKey = $orderKey->getTimeStamp();
+                if ($result !== 0) {
+                    break;
+                }
             }
-            $keyMap[] = $orderKey;
-            $valueMap[] = $value;
-        }
 
-        if ($sortType == Helper\LinqHelper::LINQ_ORDER_TYPE_DATETIME) {
-            $sortType = Helper\LinqHelper::LINQ_ORDER_TYPE_NUMERIC;
-        }
+            return $result;
+        });
+    }
 
-        if ($direction == Helper\LinqHelper::LINQ_ORDER_ASC) {
-            asort($keyMap, $sortType == Helper\LinqHelper::LINQ_ORDER_TYPE_NUMERIC ? SORT_NUMERIC : SORT_LOCALE_STRING);
-        } else {
-            arsort($keyMap, $sortType == Helper\LinqHelper::LINQ_ORDER_TYPE_NUMERIC ? SORT_NUMERIC : SORT_LOCALE_STRING);
+    public function compare($a, $b)
+    {
+        if(is_string($a) && is_string($b))
+        {
+            return strcasecmp($a, $b);
         }
-
-        $sorted = new ArrayIterator(array());
-        foreach ($keyMap as $key => $value) {
-            $sorted[] = $valueMap[$key];
+        else
+        {
+            if($a == $b) return 0;
```

Comment:
I think the typesafe operator should be use.
Whats the behavior of asort when sorting null and false etc.?
We should align with the default asort/arsort behavior and verify it with a test.
See also this table: http://stackoverflow.com/a/80649